### PR TITLE
Week 2 Dynamic Commenting

### PR DESCRIPTION
**Pull request by smitholive**
- implemented dynamic commenting in the form of a guestbook in contact.html (moved from index.html)

- Guestbook's html form prompts the user for their name and a string message.

- when comments are fetched, the name, message, and timestamp of when the comment was made are displayed

- gets and posts respond with JSON objects as opposed to directly writing html

- Java's Datastore library stores messages instead of temporarily storing them in an ArrayList

- README updated

### CODE REVIEW

File: walkthroughs/week-2-server/examples/subtraction-game/src/main/java/com/google/sps/data/SubtractionGame.java [hunk 1]
```diff
@@ -33,6 +33,7 @@
   /** Whether this game has ended, i.e. one of the players reached 0. */
   private boolean gameOver = false;
 
+
```
Comment (classm-sps) [inline walkthroughs/week-2-server/examples/subtraction-game/src/main/java/com/google/sps/data/SubtractionGame.java:36]: please revert this change, let's only make changes in the portfolio directory.

File: portfolio/src/main/webapp/index.html [hunk 2]
```diff
@@ -34,8 +34,6 @@ <h1 style:"float:right">Olivia Smith's Homepage</h1>
           Feel free to click around to learn more about me and the work I do!
       </p>
       <img src="/images/catbus.gif" alt="catbus" style="width:10vw;margin-top:10px" class="center">
-      <button onclick="getComments()" style="margin:15px">Fetch Comments</button>
```
Comment (classm-sps) [inline portfolio/src/main/webapp/index.html:37]: I haven't exhaustively checked, but you may be able to remove the `<script src="script.js"...` import now that you've removed the call from this page...

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -24,9 +34,51 @@
...
+      Date timestamp = new Date((long) entity.getProperty("timestamp"));
+      names.add(name);
+      messages.add(message);
+      timestamps.add(timestamp);
+    }
     response.setContentType("text/html;");
-    response.getWriter().println("Hello Olivia!");
+    for(int i = 0; i < messages.size(); i++){
+        response.getWriter().printf("%s at %s said: ", names.get(i), timestamps.get(i).toString());
+        response.getWriter().println(messages.get(i));
```
Comment (classm-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:62]: I think you can move this inside the results loop above, saving the need for the 3 parallel intermediate lists (and the index-based for loop).

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -24,9 +34,51 @@
...
+      String message = (String) entity.getProperty("message");
+      Date timestamp = new Date((long) entity.getProperty("timestamp"));
+      names.add(name);
+      messages.add(message);
+      timestamps.add(timestamp);
+    }
     response.setContentType("text/html;");
-    response.getWriter().println("Hello Olivia!");
+    for(int i = 0; i < messages.size(); i++){
+        response.getWriter().printf("%s at %s said: ", names.get(i), timestamps.get(i).toString());
```
Comment (classm-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:61]: might as well just combine these into a single printf:

```
response.getWriter().printf("%s at %s said: %s\n", name, timestamp, message);
```

File: portfolio/pom.xml [hunk 5]
```diff
@@ -23,6 +23,16 @@
       <version>4.0.1</version>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>com.google.code.gson</groupId>
```
Comment (classm-sps) [inline portfolio/pom.xml:27]: I see Gson being imported here, but looks like it's not being used. In fact, it sort of looks like step 3 hasn't been done, since you're passing back HTML with preformatted contents, instead of JSON that can be rendered client-side. I'd recommend that you do the JSON part, as it's a really common and important technique.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -24,9 +34,51 @@
...
+    Entity entryEntity = new Entity("Entry");
+    entryEntity.setProperty("timestamp", date.getTime());
+    entryEntity.setProperty("message", message);
+    entryEntity.setProperty("name", name);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(entryEntity);
+
+    // note: this URL will have to change once site is deployed
+    response.sendRedirect("https://8080-dot-12685145-dot-devshell.appspot.com/contact.html");
```
Comment (classm-sps) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:82]: I think you can just redirect to `/contact.html`, and it will automatically keep you on the same server.

File: portfolio/pom.xml (hunk 5 above)
Comment (smitholive) [reply portfolio/pom.xml:27]: Thank you for catching that! I think that was part of the work I accidentally lost.

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -35,7 +35,16 @@ function addRandomGreeting() {
...
+    fetch('/data').then(response => response.json()).then((commentsJson) => {
+      const commentsElement = document.getElementById('comment-container');
+      commentsElement.innerHTML = '';
+
+      // iterate through all the comment JSON objects and
+      // format them correctly
+      for(var i = 0; i < commentsJson.length; i++){
+        var comment = commentsJson[i];
+        commentsElement.innerHTML += comment.name + ' at ' 
+        + comment.timestamp + ' said: ' + comment.message + '<br>';
```
Comment (classm-sps) [inline portfolio/src/main/webapp/script.js:47]: So this comment brings up an issue that is more complicated than we need to worry about in SPS, but I at least want to make you aware of the issue. By rendering the user's comment directly into HTML, you are actually allowing an attacker (because, unfortunately, all users are also potential attackers) to inject HTML tags into your comments, which could have unexpected results:

- they could inject hyperlinks to malicious websites
- they could inject offensive images
- they could inject buttons that could execute scripts

you get the idea.

A way to avoid this would be to create the HTML elements programmatically (e.g., a <p> for each comment), and set `.innerText` for the element. Then, even if users tried to inject nasty stuff, we'd just see the text, and it wouldn't actually do anything.

I'll let you decide what you want to do about this. Let me know.
